File: UsbSerialConsole/app/src/main/java/jp/sugnakys/usbserialconsole/settings/SerialPortPreferenceFragment.java
package jp.sugnakys.usbserialconsole.settings;

import android.content.Intent;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.support.v7.widget.Toolbar;

import jp.sugnakys.usbserialconsole.R;
import jp.sugnakys.usbserialconsole.UsbService;
import jp.sugnakys.usbserialconsole.util.Log;

public class SerialPortPreferenceFragment extends BasePreferenceFragment {

    private static final String TAG = "SerialPortPreferenceFragment";

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        addPreferencesFromResource(R.xml.fragment_serial_port_preference);

        listPrefKeys = new String[]{
                getString(R.string.baudrate_key), getString(R.string.databits_key),
                getString(R.string.parity_key), getString(R.string.stopbits_key),
                getString(R.string.flowcontrol_key)};
    }

    @Override
    public void onResume() {
        super.onResume();

        Toolbar toolbar = (Toolbar) getActivity().findViewById(R.id.toolbar);
        toolbar.setTitle(getString(R.string.serial_port_title));
    }

    @Override
    public void onSharedPreferenceChanged(SharedPreferences sharedPreferences, String key) {
        super.onSharedPreferenceChanged(sharedPreferences, key);

        Log.d(TAG, "Restart UsbService");
        Intent intent = new Intent(UsbService.ACTION_SERIAL_CONFIG_CHANGED);
        getActivity().sendBroadcast(intent);
    }
}

File: UsbSerialConsole/app/src/main/java/jp/sugnakys/usbserialconsole/settings/SerialPortPreferenceFragment.kt
package jp.sugnakys.usbserialconsole.settings

import android.content.Intent
import android.os.Bundle
import androidx.preference.ListPreference
import androidx.preference.PreferenceFragmentCompat
import dagger.hilt.android.AndroidEntryPoint
import jp.sugnakys.usbserialconsole.R
import jp.sugnakys.usbserialconsole.UsbService
import timber.log.Timber

@AndroidEntryPoint
class SerialPortPreferenceFragment : PreferenceFragmentCompat() {

    override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
        setPreferencesFromResource(R.xml.fragment_serial_port_preference, rootKey)

        findPreference<ListPreference>(getString(R.string.baudrate_key))?.setOnPreferenceClickListener {
            restartService()
            false
        }

        findPreference<ListPreference>(getString(R.string.databits_key))?.setOnPreferenceClickListener {
            restartService()
            false
        }

        findPreference<ListPreference>(getString(R.string.parity_key))?.setOnPreferenceClickListener {
            restartService()
            false
        }

        findPreference<ListPreference>(getString(R.string.stopbits_key))?.setOnPreferenceClickListener {
            restartService()
            false
        }

        findPreference<ListPreference>(getString(R.string.flowcontrol_key))?.setOnPreferenceClickListener {
            restartService()
            false
        }
    }

    private fun restartService() {
        Timber.d("Restart UsbService")
        val intent = Intent(UsbService.ACTION_SERIAL_CONFIG_CHANGED)
        activity?.sendBroadcast(intent)
    }
}